Rewritten structuring.md

jlitewski opened the pull request:
Rewrote most of the article to flow better, and to show better and more robust examples.

jlitewski committed:
Rewritten structuring.md

Rewrote most of the article to flow better, and to show better and more robust examples.

bonii-xx commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,83 @@
-Generally, we'll start with a file named after your mod, and put into your package. This is the *entry point* to your mod
-and will contain some special indicators marking it as such.
+Generally, we'll start with a file named after your mod, and put into your package. This is the *entry point* to your mod and will contain some special annotations marking it as such. These annotations are `@Mod`, `@Instance`, and `@EventHandler`. Let's take a look at this example:
+
```
The example code should be in a codeblock

HACKhalo2 replied:
It should already be in a code block.

`This is a line of code that uses a tilde`

    This is a block of code
    that uses 4 spaces per line
        to show indentation and stuff

Unless there is a different one I don't know about?

bonii-xx commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,83 @@
 -------------
 
-This is an annotation indicating to the Forge Mod Loader that the class is a Mod entry point. It contains various metadata about the mod. It also designates the class that will receive `@EventHandler` events. More information can be found at... (Coming Soon)
+`@Mod` is the annotation indicating that the class is a *Mod entry point* for Forge Mod Loader. It contains various bits of Metadata that the mod will use as well as modifiers that tell Forge how to handle your mod.
```
What modifiers are you talking about here?

HACKhalo2 replied:
Yea, that was kinda sparce, huh? I'll flesh it out a bit.

bonii-xx commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,83 @@
+`@EventHandler` is an annotation used to mark whatever method it's attached to as a handler for a *Forge Lifecycle Event*. There are two main categories of Lifecycle Events: **Standard** and **Server**. **Standard Lifecycle Events** are always fired when the client and server first starts up, and **Server Lifecycle Events** are only fired when the server, either internal or dedicated, starts up.  ...
 
-!!! note
+For a simple mod, *FMLPreInitializationEvent*, *FMLInitializationEvent*, and *FMLPostInitializationEvent* should be more than enough to set up everything you need.
```
Add a link to the [loading stages](https://mcforge.readthedocs.io/en/latest/conventions/loadstages/) here.

HACKhalo2 replied:
Good Idea

bonii-xx commented:
I like it. Besides the few notes above.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,83 @@
+    public class AwesomeSauceMod
+    {
+        @Instance
+        public static AwesomeSauceMod instance = new AwesomeSauceMod();
```
No.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,83 @@
-Rather than clutter up a single class and package with everything, it is recommended you break your mod into subpackages.
+What is `@Instance`?
+-------------------
+`@Instance` is actually a child annotation of `@Mod`. It basically just gives Forge and other Mod Developers a way to instance your mod if they need to.
```
This is a horrible and inaccurate description of what @Instance does.

HACKhalo2 replied:
From the JavaDocs:

    /**
      * Populate the annotated field with the mod instance based on the specified ModId. This can be used
      * to retrieve instances of other mods.
      * @author cpw
    **/

That's what I based it on, however I'm also not entirely sure what it does anyways, so I'll look around and learn about it and rewrite this section after I educate myself.

diesieben07 replied:
It inserts the instance of the mod specified via modID in the annotation into the field. You don't populate the field yourself for a start (like you did above).

HACKhalo2 replied:
Oh, huh. I could of swore that is how you use to do it. Maybe an artifact I picked up from older versions of Forge? I've been tinkering with it since Minecraft 1.0, so it's been a log time

diesieben07 replied:
Many people do it, but it's useless and misleading, as Forge will immediately overwrite the field anyways.

shadowfacts commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,83 @@
 
-You can find an example mod in the [Forge src download](http://files.minecraftforge.net/).
+Now, the `@Mod` annotation can take quite a few arguments, but the only three *required* ones are `modid`, `name`, and `version`.
+* `modid` is the unique identifier and *internal* name of the mod, and takes a `String`. Usually, this is the same name as the mod, but in all lowercase. It's used internally by Forge and by other Mod Developers that wish to find your mod through Forge (for instance, to use your APIs). 
```
You should note the new requirements here. (It must be all lowercase and not more than 64 characters long)

HACKhalo2 replied:
I actually didn't know this. Noted.

HACKhalo2 commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,83 @@
+            // some example code
+            System.out.println("DIRT BLOCK >> "+Blocks.DIRT.getUnlocalizedName());
+        }
+    }
```
I mean, yea, I did take it from the MDK, but having at bare bones example in documentation for people that don't have the MDK configured yet and they are looking at the documentation before diving in should at least be considered. I also find it hard to follow documentation if you have to follow a link back and forth from code and the docs. I'll add a link to the example mod on the github and remove the code block if nessasary

diesieben07 replied:
What I am saying is that this code example does not provide any additional value besides what you have already explained.

HACKhalo2 replied:
Fair enough

LexManos replied:
On top of that, copy pasta, even if it is from our example mod, is bad.

tterrag1098 replied:
This needs to go before merge

jlitewski committed:
First round of changes

jlitewski committed:
Removed @Instance section

jlitewski committed:
Fixed some minor things

Lack of coffee in the morning isn't the best way to start writing documentation

LexManos commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,79 @@
+**HOWEVER**
 
-    If you do not own a domain, do not use it for your top level package. It is perfectly acceptable to start your package with anything, such as your name/nickname, or the name of the mod.
+If you do not own a domain, **do *NOT* use it for your top level package!**  
```
The second preferred option is to use your email, so people have a way of associating you with something you own.
EXA: com.gmail.lexmanos

LexManos commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,79 @@
-* Finally, a `TileEntity` for a block called `SuperChewer` would be a `tile` or `tileentity` package, with a class name of `TileSuperChewer`.
-
-Prepending your class names with what *kind* of object they are makes it easier to figure out what a class is, or guess the class for an object.
+Naming your classes, and being consistant, is important to knowing what those classes do. `BlockSauce` should be a Block, `ItemSpoon` should be an Item, et cetera. However, making sure those classes are in the right package will also help with the ability to find that class if you decide you need to look at it one day down the line. If you have `EntitySauceMonster` in the `com.examplemod.awesomes ...
```
You need to make note that these are purely suggestions and not real good ones at that. Name things whatever ytou want as long as they start with Upper case and are descriptive enough to tell you what you want to know about the class.

jlitewski committed:
Second round of changes

per @LexManos suggestions

HACKhalo2 commented:
Thank you for the suggestions guys, and @LexManos I changed the areas you pointed out.

condensedapple commented on docs/gettingstarted/structuring.md:
```diff
@@ -1,54 +1,79 @@
+* `name` is the *external* name of the mod, and takes a `String`. This is literally just the name of your mod, in our case it would be `name = "AwesomeSauce"`.
+* `version` is the version of the mod, and takes a `String`. There is no set standard for how you version things, but it's recommended that you read the [versioning](https://mcforge.readthedocs.io/en/latest/conventions/versioning/) article.
+
+Simple, right? `@Mod` has a wide range of other things you can throw at it, like the `modLanguage` example above, but those will be covered in the articles that use them.
```
I suggest to replace `modLanguage` with `modLanguageAdapter`, as it probably won't exist anymore in near future (see [here](https://github.com/MinecraftForge/MinecraftForge/pull/4029/files?diff=unified#diff-c8a307019565332befe89f9f61b2975bL161)).
